### PR TITLE
ddl: refactor and add some unit tests for label rules

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Problem Summary: Previously, the region label feature is lack of unit tests. This PR does tiny refactor and adds some basic unit tests for the region label.

### What is changed and how it works?

What's Changed:

- use a map for the return value of `GetLabelRules` 
- add truncate/rename table tests
- add drop/truncate/exchange partition tests
- add default keyword tests

### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by Yisaer — APPROVED** [review]

File: ddl/partition.go [hunk 1]
```diff
@@ -1381,14 +1381,12 @@ func (w *worker) onExchangeTablePartition(d *ddlCtx, t *meta.Meta, job *model.Jo
...
 	var ntr, ptr *label.Rule
-	for _, rule := range rules {
-		if rule.ID == ntrID {
-			ntr = rule
-		} else if rule.ID == ptrID {
-			ptr = rule
-		} else {
-			logutil.BgLogger().Error("[ddl] unexpected label rule", zap.Any("rule", rule))
-		}
+	if r, ok := rules[ntrID]; ok {
```
Comment (xhebox) [inline ddl/partition.go:1384]: `ntr, _ := rules[ntrID]` is possible

File: ddl/table.go [hunk 2]
```diff
@@ -520,31 +520,28 @@ func onTruncateTable(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ erro
...
+		newRules = append(newRules, r.Clone().Reset(newTableID, job.SchemaName, tblInfo.Name.L))
 	}
 
 	if tblInfo.GetPartitionInfo() != nil {
-		for _, r := range oldRules {
-			for _, def := range tblInfo.GetPartitionInfo().Definitions {
-				if r.ID == fmt.Sprintf(label.PartitionIDFormat, label.IDPrefix, job.SchemaName, tblInfo.Name.L, def.Name.L) {
-					newRules = append(newRules, r.Clone().Reset(def.ID, job.SchemaName, tblInfo.Name.L, def.Name.L))
-				}
+		for _, def := range tblInfo.GetPartitionInfo().Definitions {
```
Comment (xhebox) [inline ddl/table.go:542]: `for idx, def ...` thus you can `r, ok := oldRules[partRuleIDs[idx]]`

File: domain/infosync/label_manager.go [hunk 3]
```diff
@@ -83,12 +83,17 @@ func (lm *PDLabelManager) GetLabelRules(ctx context.Context, ruleIDs []string) (
...
+	for _, r := range rules {
+		ruleMap[r.ID] = r
+	}
+	return ruleMap, err
 }
 
 type mockLabelManager struct {
 	sync.RWMutex
-	labelRules map[string]*label.Rule
+	labelRules map[string][]byte
```
Comment (xhebox) [inline domain/infosync/label_manager.go:96]: There  is no need to `marshal/unmarshak`, so you want a more real mock API?

File: ddl/attributes_sql_test.go [hunk 4]
```diff
@@ -50,7 +53,13 @@ func (s *testDBSuite8) TestAlterTablePartitionAttributes(c *C) {
 	tk := testkit.NewTestKit(c, s.store)
 	tk.MustExec("use test")
 	tk.MustExec("drop table if exists t1")
-	defer tk.MustExec("drop table if exists t1")
+	defer func() {
+		tk.MustExec(`alter table t1 partition p0 attributes="";`)
+		tk.MustExec(`alter table t1 partition p1 attributes="";`)
+		tk.MustExec(`alter table t1 partition p2 attributes="";`)
+		tk.MustExec(`alter table t1 partition p3 attributes="";`)
+		tk.MustExec("drop table if exists t1")
```
Comment (xhebox) [inline ddl/attributes_sql_test.go:61]: drop table should drop attributes of partitions too, I think. We should not need to add `alter table t1 partition p3 attributes=""`, if we added, something is wrong.

**Review by xhebox — COMMENTED** [review]
Clearly not all ocurrences of `GetLabelRules` are refactored.

File: domain/infosync/label_manager.go (hunk 3 above)
Comment (rleungx) [reply domain/infosync/label_manager.go:96]: If we don't `marshal/unmarshal` it, we need to do type conversion explicitly.

File: ddl/attributes_sql_test.go (hunk 4 above)
Comment (rleungx) [reply ddl/attributes_sql_test.go:61]: The reason why I do it is to prevent the influence among cases.

**Review by xhebox — COMMENTED** [review]
`onTruncateTablePartition` should be refactored, too?

**Review by xhebox — APPROVED** [review]

**Pull request merged**